eng/packages/http-client-csharp/generator/Azure.Generator/src/Utilities/ResourceDetection.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+using Azure.Generator.Mgmt.Models;
+using Microsoft.Generator.CSharp.Input;
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace Azure.Generator.Utilities
+{
+    internal static class ResourceDetection
+    {
+        private const string ProvidersSegment = "/providers/";
+        private static ConcurrentDictionary<string, (string Name, InputModelType? InputModel)?> _resourceDataSchemaCache = new ConcurrentDictionary<string, (string Name, InputModelType? InputModel)?>();
+
+        public static bool IsResource(this OperationSet set, InputNamespace? inputNamespace = null)
+        {
+            return set.TryGetResourceDataSchema(out _, out _, inputNamespace);
+        }
+
+        private static InputModelType? FindObjectSchemaWithName(string name, InputNamespace? inputNamespace = null)
+            => inputNamespace?.Models.OfType<InputModelType>().FirstOrDefault(inputModel => inputModel.Name == name);
```

Comment:
would it be more reasonable if we access the models via `InputFactory`?

Reply:
yes, updated